src/Ocelot/LoadBalancer/LoadBalancers/CookieStickySessions.cs

```diff
@@ -1,85 +1,81 @@
+                    _loadBalancer.Release(session.HostAndPort);
                 }
-            });
+            }
         }
+    }
 
-        public async Task<Response<ServiceHostAndPort>> Lease(HttpContext httpContext)
+    public Task<Response<ServiceHostAndPort>> Lease(HttpContext httpContext)
+    {
+        var route = httpContext.Items.DownstreamRoute();
+        var serviceName = route.LoadBalancerKey;
+        var cookie = httpContext.Request.Cookies[_cookieName];
+        var key = $"{serviceName}:{cookie}"; // strong key name because of static store
+        lock (_lock)
         {
-            var key = httpContext.Request.Cookies[_key];
-
-            lock (_lock)
+            if (!string.IsNullOrEmpty(key) && _stored.TryGetValue(key, out StickySession cached))
             {
-                if (!string.IsNullOrEmpty(key) && _stored.ContainsKey(key))
-                {
-                    var cached = _stored[key];
-
-                    var updated = new StickySession(cached.HostAndPort, DateTime.UtcNow.AddMilliseconds(_keyExpiryInMs), key);
-
-                    _stored[key] = updated;
-
-                    _bus.Publish(updated, _keyExpiryInMs);
-
-                    return new OkResponse<ServiceHostAndPort>(updated.HostAndPort);
-                }
+                var updated = new StickySession(cached.HostAndPort, DateTime.UtcNow.AddMilliseconds(_keyExpiryInMs), key);
+                Update(key, updated);
+                return Task.FromResult<Response<ServiceHostAndPort>>(new OkResponse<ServiceHostAndPort>(updated.HostAndPort));
             }
 
-            var next = await _loadBalancer.Lease(httpContext);
-
+            // There is no value in the store, so lease it now!
+            var next = _loadBalancer.Lease(httpContext).GetAwaiter().GetResult(); // unfortunately the operation must be synchronous
             if (next.IsError)
             {
-                return new ErrorResponse<ServiceHostAndPort>(next.Errors);
-            }
-
-            lock (_lock)
-            {
-                if (!string.IsNullOrEmpty(key) && !_stored.ContainsKey(key))
-                {
-                    var ss = new StickySession(next.Data, DateTime.UtcNow.AddMilliseconds(_keyExpiryInMs), key);
-                    _stored[key] = ss;
-                    _bus.Publish(ss, _keyExpiryInMs);
-                }
+                return Task.FromResult<Response<ServiceHostAndPort>>(new ErrorResponse<ServiceHostAndPort>(next.Errors));
             }
 
-            return new OkResponse<ServiceHostAndPort>(next.Data);
+            var ss = new StickySession(next.Data, DateTime.UtcNow.AddMilliseconds(_keyExpiryInMs), key);
+            Update(key, ss);
+            return Task.FromResult<Response<ServiceHostAndPort>>(new OkResponse<ServiceHostAndPort>(next.Data));
         }
+    }
 
-        public void Release(ServiceHostAndPort hostAndPort)
-        {
-        }
+    protected void Update(string key, StickySession value)
```

Comment:
Use ```Lock``` and plain and simple dictionary instead, eg:
```c#
protected void Update(string key, StickySession value)
    {
        lock (Lock)
        {
            Stored[key] = value;
            _bus.Publish(value, _keyExpiryInMs);
        }
    }
```

Reply:
I've followed all your suggestions for `CookieStickySessions` and refactored.
See combined changes in https://github.com/ThreeMammals/Ocelot/pull/2111/commits/c4c7ae0ff2a891d64b4f81b6a6d051cc15e71211#diff-75ae5165d973013bcfd7ad2d5776a4868cac5f7d22931687de8fc63c80e7e36b